---
title: "Multi-Group Multi-Sample Cell-Cell Communication Analysis via NicheNet: HNSCC application"
author: "Robin Browaeys"
date: "2021-02-15"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Basic analysis of cell-cell communication from spatial data: spatial regions approach}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- github markdown built using 
rmarkdown::render("vignettes/basic_analysis.Rmd", output_format = "github_document")
-->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  # comment = "#>",
  warning = FALSE,
  message = FALSE
)
```

In this vignette, you can learn how to perform a basic spatially aware cell-cell communication analysis. 

The pipeline of a basic analysis consist mainly of the following steps:

* 1. Integration of scRNAseq with spatial transcriptomics data to spatially annotate cells
* 2. Perform cell-cell communication analysis for cells residing in same spatial regions

This vignette guides you in detail through all these steps. As example expression data of interacting cells, we will use data from XXX.

# Step 0: Load required packages and processed scRNAseq and Visium expression data 

To see how to process 10x Visium data: you can check the following vignette [Analyze visium data: clustering and pseudospace inference of spots](analyze_visium.md):`vignette("analyze_visium", package="spatialnichenetr")`

For the analysis here: we assume that the 10X Visium data contains a metadata column 'region' (and optionally 'pseudospace' if applicable), and that the scRNAseq data contains a metadata column 'celltype', which are also the identities.

Packages:

```{r}
library(spatialnichenetr)
library(nichenetr)
library(Seurat)
library(tidyverse)
```

scRNAseq data:

```{r}
seurat_obj_scrnaseq = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/scRNAseq/data_scrnaseq.rds")
DimPlot(seurat_obj_scrnaseq, reduction = "umap",pt.size = 1)
```

10X Visium spatial transcriptomics:

```{r}
seurat_obj_visium = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/visium/data_visium_regions.rds")
DimPlot(seurat_obj_visium, reduction = "umap",pt.size = 1)
SpatialDimPlot(seurat_obj_visium)
```

# Step 1: Integrate the spatial transcriptomics (10x Visium) and scRNAseq data: Define cell type composition of spots (optional)

```{r}
if(DefaultAssay(seurat_obj_visium) != "SCT"){
  DefaultAssay(seurat_obj_visium) = "SCT"
}
if(DefaultAssay(seurat_obj_scrnaseq) != "SCT"){
  DefaultAssay(seurat_obj_scrnaseq) = "SCT"
}
seurat_obj_visium = predict_celltypes_of_spots(seurat_obj_visium, seurat_obj_scrnaseq)
DefaultAssay(seurat_obj_visium) = "celltypes" # for visualization
```

Visualize the cell type prediction scores of each spot: tissue view

```{r, fig.width=11, fig.height=8}
# all cell types at once
basic_plot = SpatialFeaturePlot(seurat_obj_visium, features = rownames(seurat_obj_visium), ncol = 4, combine = FALSE, crop = TRUE, pt.size.factor = 1.66)

custom_scale_fill = scale_fill_gradientn(colours = RColorBrewer::brewer.pal(n = 8, name = "Spectral") %>% rev(),values = c(0,0.025,0.05,0.10,0.15,0.22,0.30,0.40,0.60, 1),  limits = c(0, 1))

custom_plots = lapply(basic_plot, function (x) x + custom_scale_fill + theme(legend.position = "top")) %>% patchwork::wrap_plots()
custom_plots
```


# Step 2: Integrate the spatial transcriptomics (10x Visium) and scRNAseq data: spatially annotate cells

Define spatial region of cells

```{r}
if(DefaultAssay(seurat_obj_visium) != "SCT"){
  DefaultAssay(seurat_obj_visium) = "SCT"
}
if(DefaultAssay(seurat_obj_scrnaseq) != "SCT"){
  DefaultAssay(seurat_obj_scrnaseq) = "SCT"
}
seurat_obj_scrnaseq = predict_regions_of_cells(seurat_obj_scrnaseq, seurat_obj_visium)
DefaultAssay(seurat_obj_scrnaseq) = "regions"

```

Visualize the 'region prediction scores' of each cell

```{r, fig.width=10}
basic_plot = FeaturePlot(seurat_obj_scrnaseq, features = c("central","midcentral","centralHbbMt","mid", "midportal","portalHbbIfn","portal","portaltriad"), ncol = 4, combine = FALSE)

custom_scale_fill = scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 4, name = "RdBu") %>% rev(),values = c(0,0.2, 0.4, 0.6,  1),  limits = c(0, 1))

custom_plots = lapply(basic_plot, function (x) x + custom_scale_fill + theme(legend.position = "none")) %>% patchwork::wrap_plots(ncol = 4)

custom_plots

# compare to the cell type predictions
DimPlot(seurat_obj_scrnaseq,label = T)
```


# Step 3: save new integrated objects

Save both the scRNAseq and the Visium Seurat objects after integration. These saved objects will be used as input later for the spatially aware cell-cell communication analyis.

```{r}
if(DefaultAssay(seurat_obj_visium) != "SCT"){
  DefaultAssay(seurat_obj_visium) = "SCT"
}
if(DefaultAssay(seurat_obj_scrnaseq) != "SCT"){
  DefaultAssay(seurat_obj_scrnaseq) = "SCT"
}
seurat_obj_visium %>% saveRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/visium/data_visium_integrated.rds")
seurat_obj_scrnaseq %>% saveRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/scrnaseq/data_scrnaseq_integrated.rds")
```

# Step 4: spatially aware cell-cell communication inference

## Step4A: Extract cells from region of interest

First, we will demonstrate how to extract cells that are only present in a region (or regions) of interest

As examples: we will take the central vein as one region of interest, and for another object, we will consider both the portal vein and the portaltriad.

```{r}
seurat_obj_visium = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/visium/data_visium_integrated.rds")
seurat_obj_scrnaseq = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/NicheNet_visium/data/liver/scrnaseq/data_scrnaseq_integrated.rds")
```

```{r}
seurat_obj_scrnaseq_central = get_seuratObj_region(region_oi = "central", seurat_obj_scrnaseq, spatial_annotation_assay = "regions", integration_method = "seurat", label_transfer_cutoff = 0.25)
seurat_obj_scrnaseq_portal_triad = get_seuratObj_region(region_oi = c("portal","portaltriad"), seurat_obj_scrnaseq, spatial_annotation_assay = "regions", integration_method = "seurat", label_transfer_cutoff = 0.25)
```

We will show now in a UMAP which of the original cells were kept (and compare to the scored cells above):

```{r}
central_plot_scores = FeaturePlot(seurat_obj_scrnaseq, features = c("central")) + scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 4, name = "RdBu") %>% rev(),values = c(0,0.2, 0.4, 0.6,  1),  limits = c(0, 1))
central_plot_scores

central_plot_cells = DimPlot(seurat_obj_scrnaseq_central)
central_plot_cells
```
```{r}
basic_plot = FeaturePlot(seurat_obj_scrnaseq, features = c("portal","portaltriad"), ncol = 2, combine = FALSE)
custom_scale_fill = scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 4, name = "RdBu") %>% rev(),values = c(0,0.2, 0.4, 0.6,  1),  limits = c(0, 1))
portal_plot_scores = lapply(basic_plot, function (x) x + custom_scale_fill + theme(legend.position = "none")) %>% patchwork::wrap_plots(ncol = 2)
portal_plot_scores

portal_plot_cells = DimPlot(seurat_obj_scrnaseq_portal_triad)
portal_plot_cells

```

These Seurat objects can now be used as input for classic ligand-receptor analysis methods, CellPhoneDB and NicheNet. In this way, you have now filtered the dataset to only consider cells that are located in the same spatial regions.


## Step 4B Perform Spatially-aware Cell-cell communication analysis: region-specific cell-cell communication inference and spatial NicheNet

### Step 4B.1: Extract information necessary as input for cell-cell communication analysis

In this step, we will define the following elements for each cell type of interest over all regions of interest: the number and fraction of cells of a cell type belonging to a specific region, the genes expressed in cells from that celltype/region, the genes upregulated in those cells, the ligands and receptors expressed in that region, and also the ligands and receptors upregulated in that region.

Because this might take a few minutes to run for all celltypes and all regions (because of several DE and expression calculations), we will here demonstrate how to get this information in one go for a few cell types and regions. The line of code necessary to do it for all cell types and all regions will be commented.

```{r}
# lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds"))
lr_network = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/StaticNicheNet/paper/data_nichenet/lr_network.rds")
lr_network = lr_network %>% mutate(from = convert_human_to_mouse_symbols(from), to = convert_human_to_mouse_symbols(to)) %>% drop_na() # because we have mouse data
lr_network = lr_network %>% mutate(bonafide_lr = ! database %in% c("ppi_prediction","ppi_prediction_go")) # to indicate whether a LR interaction is documented in manually annotated databases or predicted
lr_network = lr_network %>% distinct(from, to, bonafide_lr)
```

```{r}
celltypes_oi = c("hepatocyte","cholangiocyte","kupffer", "endothelial","stellate")
regions_oi = c("central", "midcentral", "mid","midportal", "portal", "portaltriad")

spatialCCC_info = get_spatialCCC_info(seurat_obj_scrnaseq = seurat_obj_scrnaseq, type = "region", spatial_annotation_assay = "regions", regions_oi = regions_oi, celltypes_oi = celltypes_oi, lr_network = lr_network, cutoff_ncells = 10, cutoff_frac_cells = 0.001, exprs_assay_oi = "SCT", exprs_frac_cutoff = 0.10)

# spatialCCC_info = get_spatialCCC_info(seurat_obj_scrnaseq = seurat_obj_scrnaseq, type = "region", spatial_annotation_assay = "regions", regions_oi = NULL, celltypes_oi = NULL, lr_network = lr_network, cutoff_ncells = 10, cutoff_frac_cells = 0.001, exprs_assay_oi = "SCT", exprs_frac_cutoff = 0.10) # this will consider all celltypes and regions of interest
```

We will now briefly show the type of information you can find in these objects:

For example: list the region-specific ligands

```{r}
spatialCCC_info[["hepatocyte"]][["central"]]$ligands_de
spatialCCC_info[["hepatocyte"]][["portal"]]$ligands_de
```

Let's now visualize some of these ligands in a FeaturePlot, and compare to the UMAP showing region predictions to demonstrate their spatially differential expression

```{r}
basic_plot = FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = "hepatocyte"), features = c("portal"))
custom_scale_fill = scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 4, name = "RdBu") %>% rev(),values = c(0,0.2, 0.4, 0.6,  1),  limits = c(0, 1))
region_plot = basic_plot + custom_scale_fill

ligand_plot = FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = "hepatocyte"), features = c("Cdh1"))

combined_plot = patchwork::wrap_plots(region_plot, ligand_plot, ncol = 2)
combined_plot
```

To get more deeper in hepatocytes around the portal vein, you could extract a specific SeuratObj containing hepatocytes only and annotating whether cells belong to the portal vein (identity = 1) or not (identity = 0)

```{r, fig.width=10}
seurat_obj_hep_portal = get_seuratObj_celltype_region(seurat_obj_scrnaseq = seurat_obj_scrnaseq, celltype_oi = "hepatocyte", region_oi = "portal")
VlnPlot(seurat_obj_hep_portal, features = "Cdh1")
DotPlot(seurat_obj_hep_portal, features = c(spatialCCC_info[["hepatocyte"]][["portal"]]$ligands_de, spatialCCC_info[["hepatocyte"]][["central"]]$ligands_de))
```


This information will allow us now to perform several types of analysis, which we will showcase now.

### Step 4B.2: get a spatially-aware ligand-receptor network

With the following command, we will construct a ligand-receptor network between all cells from the cell types in `spatialCCC_info` for each region in `spatialCCC_info` separately. Each ligand-receptor interaction will also be annotated by additional such as the spatially DE status of ligand/receptor, whether sesnder/receiver cell types are region-specific, the average expression of the ligand/receptor and product of both, scaled over the different regions,....

```{r}
spatial_lr_network = get_spatial_specific_lr_network(seurat_obj_scrnaseq, spatialCCC_info, lr_network)
head(spatial_lr_network)
nrow(spatial_lr_network)
```

As you can see, this table contains much information. Therefore we will do some prioritization to find back the most region-specific ligand-receptor pairs:
For example: in first instance, we might be most interested in a ligand-receptor interaction of which at least one of both ligand and receptor is higher expressed in the region of interest, one of both is also cell-type specific, and sender cell type is different from receiver cell type.

```{r}
spatial_lr_network_filtered = spatial_lr_network %>% filter( (ligand_spatial_de) | (receptor_spatial_de)) %>%
  filter(ligand_sender_marker | receptor_receiver_marker) %>% filter(sender != receiver) 
head(spatial_lr_network_filtered)
nrow(spatial_lr_network_filtered)
```

Apparently, there are still many interactions left, so we will filter even more deeply before prioritization, e.g. by focusing on differential interactions between two cell types in more than one region

```{r}
sender_receiver_oi = spatial_lr_network_filtered %>% ungroup() %>% distinct(sender_receiver, region) %>% group_by(sender_receiver) %>% count() %>% filter(n > 1)
sender_receiver_oi
sender_receiver_oi = sender_receiver_oi %>% pull(sender_receiver)

spatial_lr_network_filtered = spatial_lr_network_filtered %>% filter(sender_receiver %in% sender_receiver_oi)
lr_interactions_filtered = spatial_lr_network_filtered %>% pull(lr_interaction) %>% unique()
sender_receiver_filtered = spatial_lr_network_filtered %>% pull(sender_receiver) %>% unique()

```

So let's make now a figure only showing these relevant lr interactions

Let's now visualize these interactions in a dotplot (color of dots = product of quantile scaled expression of ligand and receptor, size = bona fide interaction or not)

```{r, fig.height= 8, fig.width = 21}
spatial_lr_network %>% filter(sender_receiver %in% sender_receiver_filtered & lr_interaction %in% lr_interactions_filtered)  %>% 
  ggplot(aes(lr_interaction, region)) +
  geom_point(aes(color = quantile_scaled_expression_lr_region, size = quantile_scaled_expression_lr_celltype)) + facet_grid(sender_receiver~.) +
  scale_x_discrete(position = "top") +
  scale_color_distiller(palette = "RdYlBu", direction = -1, na.value = "white") +
  theme_classic() +
  theme(axis.ticks = element_blank(), axis.title.x = element_text(size = 12), axis.text.y = element_text(face = "italic", size = 9), axis.text.x = element_text(size = 9,  angle = 90,hjust = 0))
```


```{r, fig.height= 8, fig.width = 21}
spatial_lr_network %>% filter(sender_receiver %in% sender_receiver_filtered & lr_interaction %in% lr_interactions_filtered)  %>% 
  ggplot(aes(lr_interaction, region)) +
  geom_point(aes(color = scaled_expression_lr_region, size = scaled_expression_lr_celltype)) + facet_grid(sender_receiver~.) +
  scale_x_discrete(position = "top") +
  scale_color_distiller(palette = "RdYlBu", direction = -1, na.value = "white") +
  theme_classic() +
  theme(axis.ticks = element_blank(), axis.title.x = element_text(size = 12), axis.text.y = element_text(face = "italic", size = 9), axis.text.x = element_text(size = 9,  angle = 90,hjust = 0))
```


### Step 4B.3: Spatial NicheNet: Extract ligand activities and target genes of a ligand in each receiver/region combination

Read in the NicheNet ligand-target matrix

```{r}
ligand_target_matrix = readRDS("C:/Users/rbrowaey/work/Research/NicheNet/StaticNicheNet/paper/data_nichenet/ligand_target_matrix.rds")
# ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
colnames(ligand_target_matrix) = ligand_target_matrix %>% colnames() %>% convert_human_to_mouse_symbols()
rownames(ligand_target_matrix) = ligand_target_matrix %>% rownames() %>% convert_human_to_mouse_symbols()
ligand_target_matrix = ligand_target_matrix %>% .[!is.na(rownames(ligand_target_matrix)), !is.na(colnames(ligand_target_matrix))]
```

Define activities and targets and check some ligands with high spatially-specific activity

```{r}
ligand_activities_targets_all = get_spatial_specific_ligand_activities_targets(seurat_obj_scrnaseq, spatialCCC_info, lr_network, ligand_target_matrix, target_n = 100)
ligand_activities_targets_all  %>% arrange(-pearson_specific_bg) %>% distinct(ligand, region, receiver, pearson_specific_bg, target) %>% head(25)
```

Remarkably, we see some high ligand activities for inflammatory cytokines (Il1b, Tnf) in midportal hepatocytes. In a next step we will check whether there are some midportal cells expressing these cytokines there by combining the ligand-receptor with ligand activity information.

### Step 4B.4: Combine the ligand-receptor information with the ligand activities and ligand-target links

```{r}
ligand_receptor_activities_targets = inner_join(spatial_lr_network, ligand_activities_targets_all, by = c("ligand", "receiver", "region"))
ligand_receptor_activities_targets  %>% filter(ligand_spatial_de | receptor_spatial_de | sender_region_specific | receiver_region_specific) %>% arrange(-pearson_specific_bg) %>% select(lr_interaction, ligand, receptor, sender, receiver, region, pearson_specific_bg, ligand_spatial_de, receptor_spatial_de, sender_region_specific, receiver_region_specific) %>% distinct()
```

And indeed, midportal KCs seem to express Il1b! Also interestingly is that the Il1-receptor is higher expressed in midportal hepatocytes than other regions. Altough Il1b expression in portal KCs seems to be higher than midportal, this might be an interesting predicion. 

Therefore, we will visually check this now by first checking expression of ligand, receptor and targets on the visium data, and secondly on the spatially annotated single-cell data.

```{r}
ligands_oi = c("Il1b","Il1rn")
receptors_oi = c("Il1r1")
targets_oi = ligand_receptor_activities_targets %>% filter(ligand == "Il1b", receiver == "hepatocyte" & region == "midportal") %>% pull(target) %>% unique() 
```

```{r}
SpatialFeaturePlot(seurat_obj_visium, ligands_oi)
SpatialFeaturePlot(seurat_obj_visium, receptors_oi)
SpatialFeaturePlot(seurat_obj_visium, targets_oi)

VlnPlot(seurat_obj_visium, ligands_oi)
VlnPlot(seurat_obj_visium, receptors_oi)
VlnPlot(seurat_obj_visium, targets_oi)

FeaturePlot(seurat_obj_visium, ligands_oi)
FeatureScatter(seurat_obj_visium, feature1 = ligands_oi[1], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = ligands_oi[2], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = targets_oi[1], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = targets_oi[2], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = targets_oi[3], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = targets_oi[4], feature2 = receptors_oi)
FeatureScatter(seurat_obj_visium, feature1 = targets_oi[6], feature2 = receptors_oi)

```


Expression on Visium does not seem very nice, but we can still check the spatially annotated scRNAseq data

```{r, fig.width=10}
basic_plot = FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte","kupffer")), features = c("mid", "midportal","portal"), ncol = 3, combine = FALSE, label = T)

custom_scale_fill = scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 4, name = "RdBu") %>% rev(),values = c(0,0.2, 0.4, 0.6,  1),  limits = c(0, 1))

custom_plots = lapply(basic_plot, function (x) x + custom_scale_fill + theme(legend.position = "none")) %>% patchwork::wrap_plots(ncol = 3)

custom_plots
```

```{r, fig.width=10}
FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte","kupffer")), features = c(ligands_oi))
FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte","kupffer")), features = c(receptors_oi))
FeaturePlot(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte","kupffer")), features = c(targets_oi))
```

```{r, fig.width=10}
seurat_obj_scrnaseq@meta.data$midportal = seurat_obj_scrnaseq@assays$regions["midportal",] %>% as.numeric()
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = receptors_oi, feature2 = "midportal")
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = targets_oi[1], feature2 =  "midportal")
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = targets_oi[2], feature2 =  "midportal")
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = targets_oi[3], feature2 =  "midportal")
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = targets_oi[4], feature2 =  "midportal")
FeatureScatter(seurat_obj_scrnaseq %>% subset(ident = c("hepatocyte")), feature1 = targets_oi[6], feature2 =  "midportal")

```


Not so nice :(

not so stringent enough?
